File: lib/steps/s13_audio_mix.py
from pathlib import Path
from typing import Any, Dict

from core.step_base import StepBase
from utils.ffmpeg_utils import FFmpegUtils


class StepAudioMix(StepBase):
    step_id = "s13_audio_mix"
    depends_on = ["s04_audio_separate", "s12_tts"]
    STEP_CONFIG_KEYS = ["music_volume", "ambient_volume", "tts_voice_volume", "original_voice_volume"]

    def run(self, workspace: Path, config: Dict[str, Any], job_state: Any) -> Dict[str, Any]:
        mixed_audio = workspace / "final_mixed_audio.wav"

        if config.get("ocr_only", False):
            print("[AudioMix] ocr_only mode enabled: Preserving 100% original audio stream (~0s).")
            demux_info = job_state.get_step_output("s02_demux") or {}
            audio_stream = Path(demux_info.get("audio_stream", workspace.parent / "demux" / "audio_stream.wav"))
            import shutil
            shutil.copy2(str(audio_stream), str(mixed_audio))
            return {
                "skipped": True,
                "mixed_audio": str(mixed_audio)
            }

        audio_info = job_state.get_step_output("s04_audio_separate") or {}
        tts_info = job_state.get_step_output("s12_tts") or {}

        music_path = Path(audio_info["music"])
        ambient_path = Path(audio_info["ambient"]) if "ambient" in audio_info else None
        effect_path = Path(audio_info["effect"])
        voice_path = Path(tts_info["translated_voice"])
        orig_voice_path = Path(audio_info["orig_voice"]) if "orig_voice" in audio_info else None

        probe_info = job_state.get_step_output("s01_probe") or {}
        duration = probe_info.get("duration")

        # Audio volume controls (0.0 = automatically disabled/muted)
        orig_vol = max(0.0, float(config.get("original_voice_volume", 0.05)))
        music_vol = max(0.0, float(config.get("music_volume", config.get("background_music_volume", 0.5))))
        ambient_vol = max(0.0, float(config.get("ambient_volume", 0.75)))
        tts_vol = max(0.0, float(config.get("tts_voice_volume", 1.0)))

        if tts_vol > 0.0 and tts_vol < max(music_vol, ambient_vol):
            print(f"[AudioMix] ⚠️ Warning: tts_voice_volume ({tts_vol}) is lower than music/ambient volume, TTS voice might be drowned out.")

        mixed_audio = workspace / "final_mixed_audio.wav"
        FFmpegUtils.mix_audio(
            music_path=music_path,
            voice_path=voice_path,
            output_path=mixed_audio,
            ambient_path=ambient_path,
            effect_path=effect_path,
            target_duration=duration,
            music_volume=music_vol,
            ambient_volume=ambient_vol,
            voice_volume=tts_vol,
            orig_voice_path=orig_voice_path,
            orig_voice_volume=orig_vol
        )

        return {
            "mixed_audio": str(mixed_audio)
        }

File: video_engine/lib/steps/s13_audio_mix.dart
import 'dart:io';
import 'package:path/path.dart' as p;
import '../core/job_state.dart';
import '../core/step_base.dart';
import '../utils/ffmpeg_utils.dart';

class StepAudioMix extends StepBase {
  @override
  String get stepId => 's13_audio_mix';

  @override
  List<String> get dependsOn => const ['s04_audio_separate', 's12_tts'];

  @override
  List<String> get stepConfigKeys => const [
        'tts_voice',
        'music_volume',
        'ambient_volume',
        'tts_voice_volume',
        'original_voice_volume',
      ];

  @override
  Future<Map<String, dynamic>> run(
    Directory workspace,
    Map<String, dynamic> config,
    JobState jobState,
  ) async {
    final mixedAudio = File(p.join(workspace.path, 'final_mixed_audio.wav'));

    final audioCfg = config['audio'] is Map ? (config['audio'] as Map) : {};
    final audioVols = audioCfg['volumes'] is Map ? (audioCfg['volumes'] as Map) : {};
    final ttsCfg = config['tts'] is Map ? (config['tts'] as Map) : {};

    final ttsVol = double.tryParse((config['tts_voice_volume'] ?? audioVols['tts_voice'])?.toString() ?? '1.0') ?? 1.0;
    final ttsVoice = (config['tts_voice'] ?? ttsCfg['voice'])?.toString().toLowerCase().trim() ?? 'vi';
    final isOcrOnly = config['ocr_only'] == true || config['ocr_only']?.toString().toLowerCase() == 'true';

    if (isOcrOnly || ttsVol == 0.0 || ttsVoice == '0' || ttsVoice == 'none' || ttsVoice == 'off') {
      final demuxInfo = jobState.getStepOutput('s02_demux') ?? {};
      final audioStream = demuxInfo['audio_stream'] as String?;
      if (audioStream != null && File(audioStream).existsSync()) {
        File(audioStream).copySync(mixedAudio.path);
      }
      return {
        'skipped': true,
        'mixed_audio': mixedAudio.path,
      };
    }

    final audioInfo = jobState.getStepOutput('s04_audio_separate') ?? {};
    final ttsInfo = jobState.getStepOutput('s12_tts') ?? {};
    final probeInfo = jobState.getStepOutput('s01_probe') ?? {};

    final musicPath = audioInfo['music'] as String? ?? '';
    final voicePath = ttsInfo['translated_voice'] as String? ?? '';
    final ambientPath = audioInfo['ambient'] as String?;
    final origVoicePath = audioInfo['orig_voice'] as String?;
    final duration = (probeInfo['duration'] as num?)?.toDouble();

    final origVol = double.tryParse((config['original_voice_volume'] ?? audioVols['original_voice'])?.toString() ?? '0.05') ?? 0.05;
    final musicVol = double.tryParse((config['music_volume'] ?? audioVols['music'])?.toString() ?? '0.5') ?? 0.5;
    final ambientVol = double.tryParse((config['ambient_volume'] ?? audioVols['ambient'])?.toString() ?? '0.75') ?? 0.75;

    await FFmpegUtils.mixAudio(
      musicPath: musicPath,
      voicePath: voicePath,
      outputPath: mixedAudio.path,
      ambientPath: ambientPath,
      origVoicePath: origVoicePath,
      targetDuration: duration,
      musicVolume: musicVol,
      ambientVolume: ambientVol,
      voiceVolume: ttsVol,
      origVoiceVolume: origVol,
    );

    return {
      'mixed_audio': mixedAudio.path,
    };
  }
}
